Make query compiler base class a Python ABC

datapythonista opened the pull request:
I was having a look at the query compiler, and thought it could be a good idea to use a Python ABC. So an exception is raised if children is missing one of the methods, and it's also more readable.

After changing that, I realized and made couple of adjustments:
- `isnull` and `notnull` wheren't being implemented, I made them aliases of `isna` and `notna` in the base class
- `to_datetime` was listed as a reduction. I think pandas provides just as a helper, and it's not part of the dataframe API. Removed it here.
- `squeeze` was not implemented in the pandas query compiler, I made it explicit in the pandas compiler itself.
- `binary_op` is implemented in the base class, but not in the pandas query compiler, where individual methods are defined. I added a comment, and make the method call the corresponding method (`binary_op(add, ...)` -> `add(...)`).

Not sure if all those make sense or are good ideas, happy to have feedback and comments.

devin-petersohn commented on modin/backends/base/query_compiler.py:
```diff
@@ -143,11 +150,15 @@ def binary_op(self, op, other, **kwargs):
-        raise NotImplementedError(NOT_IMPLEMENTED_MESSAGE)
+        # TODO This doesn't seem to be used, children classes seem
+        # to implement the binary methods directly
+        return getattr(self, op)(other, **kwargs)
```
Instead of `binary_op` let's enumerate them in the base class as well. The list can be found here: https://github.com/modin-project/modin/blob/9e21b589d0baaee3de4f31fe2171f3dfaa9363a7/modin/backends/pandas/query_compiler.py#L194-L220

devin-petersohn commented on modin/backends/pandas/query_compiler.py:
```diff
@@ -1377,6 +1377,9 @@ def view(self, index=None, columns=None):
             self._modin_frame.mask(row_numeric_idx=index, col_numeric_idx=columns)
         )
 
+    def squeeze(self, ndim=0, axis=None):
```
`squeeze` can be deleted here and in the parent class, it is not used.

devin-petersohn commented on modin/backends/base/query_compiler.py:
```diff
@@ -213,88 +228,102 @@ def reset_index(self, **kwargs):
-        raise NotImplementedError(NOT_IMPLEMENTED_MESSAGE)
+        pass
 
     def notnull(self):
```
`notnull` can be deleted we want to only expose one way to do `notna`.

`notnull = notna` in the API layer: https://github.com/modin-project/modin/blob/9e21b589d0baaee3de4f31fe2171f3dfaa9363a7/modin/pandas/base.py#L1756

devin-petersohn commented on modin/backends/base/query_compiler.py:
```diff
@@ -213,88 +228,102 @@ def reset_index(self, **kwargs):
-        raise NotImplementedError(NOT_IMPLEMENTED_MESSAGE)
+        pass
 
     def isnull(self):
```
`isnull` can be deleted.

`isnull = isna` in the API layer:
https://github.com/modin-project/modin/blob/9e21b589d0baaee3de4f31fe2171f3dfaa9363a7/modin/pandas/base.py#L1469

datapythonista commented:
Thanks for all the feedback, addressed all the comments, and squashed into a single signed-off commit.

Not sure about coverage, not sure if tools are smart enough to know that abstract methods code are not reachable. But I'd assume the percentage of the `pass` lines is very small, and shouldn't have an impact on the coverage ratio.

devin-petersohn commented:
Thanks @datapythonista!
